File: StackStackStack/StackStackStack/Library/LibraryStore.swift
import Foundation
import StackEngineCore

/// Library mutation failures surfaced to the editor UI.
enum LibraryError: LocalizedError {
    case recordMissing
    var errorDescription: String? { "This stack no longer exists in the library." }
}

/// Minimal file-backed library: results (JPEG or HEIC) + a JSON index in the given root. All writes
/// are atomic (temp-file + rename) and data-protected, and `loadAll` self-heals (drops records whose
/// file vanished) and preserves a corrupt index rather than letting the next save overwrite it.
///
/// `@unchecked Sendable`: no mutable in-memory state — all stored properties are immutable and every
/// method is stateless file I/O. Reads (loadAll/originalData/adjustments) are safe to call off the
/// main thread. WRITES (save/applyEdit/delete) must stay MainActor-confined: they are read-modify-
/// write on index.json, so two concurrent writers could lose a record (Sendable guarantees memory
/// safety of the reference, not mutation atomicity).
final class LibraryStore: @unchecked Sendable {
    private let root: URL
    private let indexURL: URL
    private let fm = FileManager.default
    /// Atomic so an interrupted write can't truncate the file; complete protection so contents are
    /// unreadable while the device is locked.
    private static let writeOptions: Data.WritingOptions = [.atomic, .completeFileProtection]

    init(rootDirectory: URL = LibraryStore.defaultRoot()) {
        self.root = rootDirectory
        self.indexURL = rootDirectory.appendingPathComponent("index.json")
        try? fm.createDirectory(at: rootDirectory, withIntermediateDirectories: true)
    }

    static func defaultRoot() -> URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Stacks", isDirectory: true)
    }

    struct SavedStack { let id: UUID; let resultURL: URL }

    @discardableResult
    func save(result: Data, reference: Data? = nil, format: ImageEncoder.Format, mode: String, frameCount: Int) throws -> SavedStack {
        let id = UUID()
        let fileName = "\(id.uuidString).\(format.fileExtension)"
        let url = root.appendingPathComponent(fileName)
        let now = Date()
        // Write the files first, then the index last — so a failure never leaves an index entry
        // pointing at a file that isn't there.
        try result.write(to: url, options: Self.writeOptions)
        try result.write(to: originalURL(for: id, format: format), options: Self.writeOptions)   // immutable original
        // The aligned reference frame — the blend-strength lerp's second endpoint; absent for depth
        // and legacy records (the slider is hidden when nil).
        if let reference {
            try reference.write(to: referenceURL(for: id, format: format), options: Self.writeOptions)
        }
        var records = (try? loadRaw()) ?? []
        records.insert(StackRecord(id: id, createdAt: now, mode: mode, frameCount: frameCount,
                                   resultFileName: fileName, updatedAt: now, format: format.rawValue), at: 0)
        try persist(records)
        return SavedStack(id: id, resultURL: url)
    }

    /// All persisted records for DISPLAY: self-heals by dropping records whose result file is gone.
    /// Mutations (save/delete/applyEdit) use `loadRaw` instead, so a transient read-time miss can
    /// never feed back into a `persist` and permanently drop a still-recoverable record.
    func loadAll() throws -> [StackRecord] {
        try loadRaw().filter { fm.fileExists(atPath: resultURL(for: $0).path) }
    }

    /// The raw decoded index, unfiltered. On a corrupt/torn index, the bytes are preserved aside
    /// (timestamped, so an earlier recoverable snapshot isn't clobbered) and [] returned — rather
    /// than letting the next save overwrite and permanently drop every prior record.
    private func loadRaw() throws -> [StackRecord] {
        guard let data = try? Data(contentsOf: indexURL) else { return [] }
        do {
            return try JSONDecoder().decode([StackRecord].self, from: data)
        } catch {
            let aside = root.appendingPathComponent("index.\(Int(Date().timeIntervalSince1970)).corrupt")
            try? fm.moveItem(at: indexURL, to: aside)
            return []
        }
    }

    func resultURL(for record: StackRecord) -> URL { record.resultURL(in: root) }

    /// The persisted record for an id (nil if absent).
    func record(for id: UUID) -> StackRecord? {
        ((try? loadRaw()) ?? []).first { $0.id == id }
    }

    /// Remove a stack's record and all of its files (result, original, reference, edit sidecar).
    func delete(id: UUID) throws {
        var records = (try? loadRaw()) ?? []
        let format = records.first(where: { $0.id == id })?.encoderFormat
        var urls = [editsURL(for: id)]
        for f in format.map({ [$0] }) ?? [.jpeg, .heic] {   // unknown record → sweep both extensions
            urls.append(resultURL(for: id, format: f))
            urls.append(originalURL(for: id, format: f))
            urls.append(referenceURL(for: id, format: f))
        }
        for url in urls { try? fm.removeItem(at: url) }
        records.removeAll { $0.id == id }
        try persist(records)
    }

    /// Delete every stack (Settings ▸ Storage). MainActor-only, like all writes.
    func deleteAll() throws {
        let records = (try? loadRaw()) ?? []
        for rec in records {
            var urls = [editsURL(for: rec.id)]
            for f in [rec.encoderFormat] as [ImageEncoder.Format] {
                urls.append(resultURL(for: rec.id, format: f))
                urls.append(originalURL(for: rec.id, format: f))
                urls.append(referenceURL(for: rec.id, format: f))
            }
            for url in urls { try? fm.removeItem(at: url) }
        }
        try persist([])
    }

    /// Total bytes of all library files (results, originals, sidecars, index). Stateless file I/O —
    /// safe to call off the main thread (Settings computes it in a background task).
    func storageUsedBytes() -> Int64 {
        guard let files = try? fm.contentsOfDirectory(at: root, includingPropertiesForKeys: [.fileSizeKey]) else { return 0 }
        return files.reduce(0) { sum, url in
            sum + Int64((try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
        }
    }

    /// Delete `<uuid>.*` files that have no matching index record (orphans from failed/partial saves).
    /// MainActor-only: must not interleave with `save` (which writes files before the index entry).
    func reconcileOrphans() {
        let ids = Set(((try? loadRaw()) ?? []).map { $0.id.uuidString })   // raw → keep files for records present
        guard let files = try? fm.contentsOfDirectory(at: root, includingPropertiesForKeys: nil) else { return }
        for f in files {
            let name = f.lastPathComponent
            // Only consider per-stack files; index.json and index.<epoch>.corrupt end in .json/.corrupt
            // (the former is excluded by name, the latter doesn't match these suffixes).
            guard name.hasSuffix(".jpg") || name.hasSuffix(".heic") || name.hasSuffix(".json"),
                  name != "index.json" else { continue }
            let uuidPart = String(name.prefix(36))   // "<uuid>.jpg" / "<uuid>.orig.jpg" / "<uuid>.edits.json" / "<uuid>.heic"
            // Non-UUID-named files are never per-stack files and must not be swept (e.g. index.corrupt).
            guard UUID(uuidString: uuidPart) != nil else { continue }
            if !ids.contains(uuidPart) { try? fm.removeItem(at: f) }
        }
    }

    private func resultURL(for id: UUID, format: ImageEncoder.Format) -> URL {
        root.appendingPathComponent("\(id.uuidString).\(format.fileExtension)")
    }
    private func originalURL(for id: UUID, format: ImageEncoder.Format) -> URL {
        root.appendingPathComponent("\(id.uuidString).orig.\(format.fileExtension)")
    }
    /// The aligned reference frame used for blend-strength editing (`<uuid>.ref.<ext>`).
    private func referenceURL(for id: UUID, format: ImageEncoder.Format) -> URL {
        root.appendingPathComponent("\(id.uuidString).ref.\(format.fileExtension)")
    }
    private func editsURL(for id: UUID) -> URL { root.appendingPathComponent("\(id.uuidString).edits.json") }

    /// The immutable original stacked image (the record's own format), used as the editing source.
    func originalData(for id: UUID) -> Data? {
        guard let rec = record(for: id) else { return nil }
        return try? Data(contentsOf: originalURL(for: id, format: rec.encoderFormat))
    }

    /// The aligned reference frame (the blend-strength lerp's second endpoint), or nil for depth
    /// records, legacy records, or any record whose ref file is absent or corrupt.
    func referenceData(for id: UUID) -> Data? {
        guard let rec = record(for: id) else { return nil }
        return try? Data(contentsOf: referenceURL(for: id, format: rec.encoderFormat))
    }

    /// The persisted adjustments for a record (identity if none / unreadable).
    func adjustments(for id: UUID) -> ImageAdjustments {
        guard let data = try? Data(contentsOf: editsURL(for: id)),
              let adj = try? JSONDecoder().decode(ImageAdjustments.self, from: data) else { return .identity }
        return adj
    }

    /// Overwrite the displayed result with a rendered image (in the record's own format), persist
    /// the adjustments, and bump `updatedAt` so gallery cells reload (their task id includes it).
    func applyEdit(id: UUID, adjustments: ImageAdjustments, rendered: Data) throws {
        var records = try loadRaw()
        guard let i = records.firstIndex(where: { $0.id == id }) else { throw LibraryError.recordMissing }
        let rec = records[i]
        try rendered.write(to: resultURL(for: id, format: rec.encoderFormat), options: Self.writeOptions)
        try JSONEncoder().encode(adjustments).write(to: editsURL(for: id), options: Self.writeOptions)
        records[i].updatedAt = Date()
        try persist(records)
    }

    private func persist(_ records: [StackRecord]) throws {
        let data = try JSONEncoder().encode(records)
        try data.write(to: indexURL, options: Self.writeOptions)
    }
}

File: android/app/src/main/kotlin/com/jdpneto/stackstackstack/LibraryStore.kt
package com.jdpneto.stackstackstack

import com.jdpneto.stackengine.ImageAdjustments
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/**
 * Minimal file-backed library: results (JPEG or HEIC) + a JSON index in [root].
 *
 * All writes are atomic (write-to-temp + rename) and `loadAll` self-heals (drops records whose
 * file vanished) and preserves a corrupt index rather than letting the next save overwrite it.
 *
 * Threading contract (mirrors iOS):
 * - **Reads** (`loadAll`, `originalData`, `referenceData`, `adjustments`, `storageUsedBytes`)
 *   are safe to call off the main thread.
 * - **Writes** (`save`, `applyEdit`, `delete`, `deleteAll`, `reconcileOrphans`) must stay
 *   on the main thread (or a single serialized thread): they are read-modify-write on
 *   `index.json`, so two concurrent writers could lose a record.
 */
class LibraryStore(val root: File = defaultRoot()) {

    private val indexFile = File(root, "index.json")

    /** kotlinx.serialization Json instance — ignores unknown keys for back-compat. */
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true   // ensure nullable fields are always written (so re-reads are stable)
    }

    init {
        root.mkdirs()
    }

    companion object {
        /**
         * Default root: `<filesDir>/Stacks`. Callers on the app side pass
         * `context.filesDir` at construction; tests inject a temp dir.
         */
        fun defaultRoot(filesDir: File = File(System.getProperty("user.home")!!)): File =
            File(filesDir, "Stacks")
    }

    // MARK: - Saved-stack handle

    data class SavedStack(val id: UUID, val resultFile: File)

    // MARK: - Write operations

    /**
     * Save a stacked result. Writes the result + original + optional reference first, then
     * updates the index last — so a failure never leaves an index entry without its file.
     *
     * @param result          Encoded image bytes (JPEG or HEIC).
     * @param reference       Aligned reference frame for blend-strength editing; null for depth
     *                        and any look that produced no reference (slider hidden when null).
     *                        A failed reference write must not lose the shot — the record simply
     *                        has no blend. (spec §6)
     * @param format          Encoder format.
     * @param mode            [com.jdpneto.stackengine.StackMode.storageKey] of the look.
     * @param frameCount      Number of burst frames stacked.
     * @param iso             First-frame ISO speed; null for legacy/fallback path.
     * @param shutterSeconds  First-frame exposure time; null for legacy/fallback path.
     */
    @Throws(Exception::class)
    fun save(
        result: ByteArray,
        reference: ByteArray? = null,
        format: ImageEncoder.Format,
        mode: String,
        frameCount: Int,
        iso: Double? = null,
        shutterSeconds: Double? = null
    ): SavedStack {
        val id = UUID.randomUUID()
        val fileName = "${id.toString().uppercase()}.${format.fileExtension}"
        val resultFile = File(root, fileName)
        val nowPosix = System.currentTimeMillis() / 1000.0
        val nowApple = StackRecord.posixToAppleEpoch(nowPosix)

        // Write files BEFORE updating the index (atomically) — failure leaves no dangling entry.
        resultFile.writeAtomically(result)
        originalFile(id, format).writeAtomically(result)       // immutable original
        if (reference != null) {
            runCatching { referenceFile(id, format).writeAtomically(reference) }
            // Swallow failures — a missing reference just hides the blend slider (spec §6).
        }

        val records = loadRaw().toMutableList()
        records.add(0, StackRecord(
            id                  = id,
            createdAtAppleEpoch = nowApple,
            mode                = mode,
            frameCount          = frameCount,
            resultFileName      = fileName,
            updatedAtAppleEpoch = nowApple,
            format              = format.rawValue,
            iso                 = iso,
            shutterSeconds      = shutterSeconds
        ))
        persist(records)
        return SavedStack(id = id, resultFile = resultFile)
    }

    /**
     * All persisted records for DISPLAY: self-heals by dropping records whose result file is gone.
     * Mutations (`save`/`delete`/`applyEdit`) use `loadRaw` instead, so a transient read-time
     * miss can never feed back into `persist` and permanently drop a still-recoverable record.
     */
    @Throws(Exception::class)
    fun loadAll(): List<StackRecord> =
        loadRaw().filter { File(root, it.resultFileName).exists() }

    /**
     * The persisted record for an id (null if absent).
     */
    fun record(id: UUID): StackRecord? =
        loadRaw().firstOrNull { it.id == id }

    /**
     * Remove a stack's record and all of its files (result, original, reference, edit sidecar).
     */
    @Throws(Exception::class)
    fun delete(id: UUID) {
        val records = loadRaw().toMutableList()
        val format = records.firstOrNull { it.id == id }?.encoderFormat
        val toDelete = mutableListOf(editsFile(id))
        // Unknown record → sweep both extensions (mirrors iOS).
        val formats: List<ImageEncoder.Format> = format?.let { listOf(it) }
            ?: listOf(ImageEncoder.Format.JPEG, ImageEncoder.Format.HEIC)
        for (f in formats) {
            toDelete += resultFile(id, f)
            toDelete += originalFile(id, f)
            toDelete += referenceFile(id, f)
        }
        toDelete.forEach { runCatching { it.delete() } }
        records.removeAll { it.id == id }
        persist(records)
    }

    /**
     * Delete every stack (Settings ▸ Storage). Main-thread only, like all writes.
     */
    @Throws(Exception::class)
    fun deleteAll() {
        val records = loadRaw()
        for (rec in records) {
            val toDelete = mutableListOf(editsFile(rec.id))
            toDelete += resultFile(rec.id, rec.encoderFormat)
            toDelete += originalFile(rec.id, rec.encoderFormat)
            toDelete += referenceFile(rec.id, rec.encoderFormat)
            toDelete.forEach { runCatching { it.delete() } }
        }
        persist(emptyList())
    }

    /**
     * Total bytes of all library files (results, originals, sidecars, index).
     * Stateless file I/O — safe to call off the main thread.
     */
    fun storageUsedBytes(): Long =
        root.listFiles()?.sumOf { it.length() } ?: 0L

    /**
     * Delete `<uuid>.*` files that have no matching index record (orphans from failed/partial saves).
     * Main-thread only: must not interleave with `save`.
     */
    fun reconcileOrphans() {
        val ids = loadRaw().map { it.id.toString().uppercase() }.toSet()
        val files = root.listFiles() ?: return
        for (f in files) {
            val name = f.name
            // Only consider per-stack files; index.json and index.<epoch>.corrupt end in .json/.corrupt.
            if (!name.endsWith(".jpg") && !name.endsWith(".heic") && !name.endsWith(".json")) continue
            if (name == "index.json") continue
            val uuidPart = name.take(36)
            // Non-UUID-named files (e.g. index.corrupt backups) are never swept.
            try { UUID.fromString(uuidPart) } catch (e: Exception) { continue }
            if (!ids.contains(uuidPart.uppercase())) {
                runCatching { f.delete() }
            }
        }
    }

    // MARK: - Read operations (off-thread-safe)

    /**
     * The immutable original stacked image (the record's own format), used as the editing source.
     */
    fun originalData(id: UUID): ByteArray? {
        val rec = record(id) ?: return null
        return runCatching { originalFile(id, rec.encoderFormat).readBytes() }.getOrNull()
    }

    /**
     * The aligned reference frame (the blend-strength lerp's second endpoint), or null for depth
     * records, legacy records, or any record whose ref file is absent or corrupt.
     */
    fun referenceData(id: UUID): ByteArray? {
        val rec = record(id) ?: return null
        return runCatching { referenceFile(id, rec.encoderFormat).readBytes() }.getOrNull()
    }

    /**
     * The persisted adjustments for a record (identity if none / unreadable).
     * If the decoded adjustments carry a blendStrength < 1 but no reference file exists, the α is
     * normalized to 1: a persisted α without its reference would silently re-bake at a different
     * look on the next save — normalize instead. (Fix 4, spec §3)
     */
    fun adjustments(id: UUID): ImageAdjustments {
        val data = runCatching { editsFile(id).readBytes() }.getOrNull()
            ?: return ImageAdjustments.identity
        val dto = runCatching {
            json.decodeFromString<ImageAdjustmentsDto>(String(data))
        }.getOrNull() ?: return ImageAdjustments.identity
        var adj = dto.toImageAdjustments()
        if (adj.blendStrength < 1f) {
            val rec = record(id)
            if (rec != null) {
                val refFile = referenceFile(id, rec.encoderFormat)
                if (!refFile.exists()) {
                    adj.blendStrength = 1f
                }
            }
        }
        return adj
    }

    /**
     * Overwrite the displayed result with a rendered image (in the record's own format), persist
     * the adjustments, and bump `updatedAt` so gallery cells reload.
     */
    @Throws(Exception::class)
    fun applyEdit(id: UUID, adjustments: ImageAdjustments, rendered: ByteArray) {
        val records = loadRaw().toMutableList()
        val i = records.indexOfFirst { it.id == id }
        if (i < 0) throw LibraryError.RecordMissing
        val rec = records[i]
        resultFile(id, rec.encoderFormat).writeAtomically(rendered)
        val dtoJson = json.encodeToString(ImageAdjustmentsDto.from(adjustments))
        editsFile(id).writeAtomically(dtoJson.toByteArray())
        val nowApple = StackRecord.posixToAppleEpoch(System.currentTimeMillis() / 1000.0)
        records[i] = rec.copy(updatedAtAppleEpoch = nowApple)
        persist(records)
    }

    // MARK: - File paths

    fun resultURL(record: StackRecord): File = File(root, record.resultFileName)

    private fun resultFile(id: UUID, format: ImageEncoder.Format): File =
        File(root, "${id.toString().uppercase()}.${format.fileExtension}")

    private fun originalFile(id: UUID, format: ImageEncoder.Format): File =
        File(root, "${id.toString().uppercase()}.orig.${format.fileExtension}")

    /** The aligned reference frame used for blend-strength editing (`<uuid>.ref.<ext>`). */
    private fun referenceFile(id: UUID, format: ImageEncoder.Format): File =
        File(root, "${id.toString().uppercase()}.ref.${format.fileExtension}")

    private fun editsFile(id: UUID): File =
        File(root, "${id.toString().uppercase()}.edits.json")

    // MARK: - Index persistence

    /**
     * The raw decoded index, unfiltered. On a corrupt/torn index, the bytes are preserved aside
     * (timestamped) and an empty list is returned — rather than letting the next save overwrite
     * and permanently drop every prior record.
     */
    private fun loadRaw(): List<StackRecord> {
        if (!indexFile.exists()) return emptyList()
        val bytes = runCatching { indexFile.readBytes() }.getOrNull() ?: return emptyList()
        return try {
            json.decodeFromString<List<StackRecord>>(String(bytes))
        } catch (e: Exception) {
            val asideName = "index.${System.currentTimeMillis() / 1000}.corrupt"
            runCatching { indexFile.renameTo(File(root, asideName)) }
            emptyList()
        }
    }

    private fun persist(records: List<StackRecord>) {
        val text = json.encodeToString(records)
        indexFile.writeAtomically(text.toByteArray())
    }
}

// MARK: - Extension helpers

/**
 * Atomic write: write to a temp file then rename over the target.
 * Prevents truncated writes from being observable. Mirrors iOS `.atomic` write option.
 */
private fun File.writeAtomically(bytes: ByteArray) {
    val tmp = File(parentFile ?: File("."), "${name}.tmp")
    tmp.writeBytes(bytes)
    tmp.renameTo(this)
}
